sumpy/p2p.py

```diff
@@ -52,76 +51,128 @@
-                name, device)
+        from pytools import single_valued
+        from sumpy.kernel import (TargetDerivativeRemover,
+                SourceDerivativeRemover)
+        tdr = TargetDerivativeRemover()
+        sdr = SourceDerivativeRemover()
+
+        if source_kernels is None:
+            source_kernels = [single_valued(tdr(knl) for knl in target_kernels)]
+            target_kernels = [sdr(knl) for knl in target_kernels]
+        else:
+            for knl in source_kernels:
+                assert(tdr(knl) == knl)
+            for knl in target_kernels:
+                assert(sdr(knl) == knl)
+
+        base_source_kernel = single_valued(sdr(knl) for knl in source_kernels)
+        base_target_kernel = single_valued(tdr(knl) for knl in target_kernels)
+        assert base_source_kernel == base_target_kernel
+
+        KernelComputation.__init__(self, ctx=ctx, target_kernels=target_kernels,
+            source_kernels=source_kernels, strength_usage=strength_usage,
+            value_dtypes=value_dtypes, name=name, device=device)
 
         self.exclude_self = exclude_self
+        self.source_kernels = source_kernels
+        self.target_kernels = target_kernels
 
-        from pytools import single_valued
-        self.dim = single_valued(knl.dim for knl in self.kernels)
+        self.dim = single_valued(knl.dim for knl in
+            list(self.target_kernels) + list(self.source_kernels))
 
     def get_cache_key(self):
-        return (type(self).__name__, tuple(self.kernels), self.exclude_self,
-                tuple(self.strength_usage), tuple(self.value_dtypes))
+        return (type(self).__name__, tuple(self.target_kernels), self.exclude_self,
+                tuple(self.strength_usage), tuple(self.value_dtypes),
+                tuple(self.source_kernels))
 
     def get_loopy_insns_and_result_names(self):
         from sumpy.symbolic import make_sym_vector
+        from pymbolic import var
+
         dvec = make_sym_vector("d", self.dim)
 
         from sumpy.assignment_collection import SymbolicAssignmentCollection
         sac = SymbolicAssignmentCollection()
 
-        result_names = [
-                sac.assign_unique("knl%d" % i,
-                    knl.postprocess_at_target(
-                        knl.postprocess_at_source(
-                            knl.get_expression(dvec),
+        isrc_sym = var("isrc")
+
+        if len(self.source_kernels) == 1:
+            in_knl = self.source_kernels[0]
+            exprs = []
+            for i, out_knl in enumerate(self.target_kernels):
+                expr = out_knl.postprocess_at_target(
+                        in_knl.postprocess_at_source(
+                            in_knl.get_expression(dvec),
                             dvec),
                         dvec)
-                    )
-                for i, knl in enumerate(self.kernels)]
+                expr *= self.get_strength_or_not(isrc_sym, 0)
+                exprs.append(expr)
```

Comment:
How come this must be treated specially? (Aka I would much prefer if this branch didn't exist.)